### PR TITLE
validate tags on copy object and add regex for validating tags

### PR DESCRIPTION
**Pull request by lapshin-vitaly**
# What problem are we solving?

1. When copy object, tags not validated. I can put some big input in tags with replace flag
2. Input symbols in tags don't validate with any regex.

# How are we solving the problem?

1. Refactor validating tags and add to separate function. 
2. Add regex for validating. Validating rules describe on amazon docs: https://docs.aws.amazon.com/directoryservice/latest/devguide/API_Tag.html

### CODE REVIEW

**Comment by kmlebedev** [comment]
@lapshin-vitaly test failed

```
--- FAIL: TestProcessMetadataBytes (0.00s)
panic: assignment to entry in nil map [recovered]
	panic: assignment to entry in nil map
goroutine 231 [running]:
testing.tRunner.func1.2({0x21b71e0, 0x2aeaf10})
	/opt/hostedtoolcache/go/1.18.3/x64/src/testing/testing.go:1389 +0x24e
testing.tRunner.func1()
	/opt/hostedtoolcache/go/1.18.3/x64/src/testing/testing.go:1392 +0x39f
panic({0x21b71e0, 0x2aeaf10})
	/opt/hostedtoolcache/go/1.18.3/x64/src/runtime/panic.go:838 +0x207
github.com/chrislusf/seaweedfs/weed/s3api.parseTagsHeader({0x25a244b?, 0x2590e3f?})
	/home/runner/work/seaweedfs/seaweedfs/weed/s3api/tags.go:49 +0x10e
github.com/chrislusf/seaweedfs/weed/s3api.processMetadataBytes(0xc00008bdb0?, 0xc00008bcd0?, 0x0, 0x1)
	/home/runner/work/seaweedfs/seaweedfs/weed/s3api/s3api_object_copy_handlers.go:286 +0x37d
github.com/chrislusf/seaweedfs/weed/s3api.TestProcessMetadataBytes(0xc0002f61a0)
	/home/runner/work/seaweedfs/seaweedfs/weed/s3api/s3api_object_copy_handlers_test.go:370 +0x19c
testing.tRunner(0xc0002f61a0, 0x265abc0)
	/opt/hostedtoolcache/go/1.18.3/x64/src/testing/testing.go:1439 +0x102
created by testing.(*T).Run
	/opt/hostedtoolcache/go/1.18.3/x64/src/testing/testing.go:1486 +0x35f
FAIL	github.com/chrislusf/seaweedfs/weed/s3api	0.038s
```

See https://github.com/chrislusf/seaweedfs/runs/6963408858?check_suite_focus=true#step:6:630

File: weed/s3api/s3api_object_copy_handlers.go [hunk 1]
```diff
@@ -45,7 +45,11 @@ func (s3a *S3ApiServer) CopyObjectHandler(w http.ResponseWriter, r *http.Request
 			s3err.WriteErrorResponse(w, r, s3err.ErrInvalidCopySource)
 			return
 		}
-		entry.Extended = processMetadataBytes(r.Header, entry.Extended, replaceMeta, replaceTagging)
+		entry.Extended, err = processMetadataBytes(r.Header, entry.Extended, replaceMeta, replaceTagging)
+		if err != nil {
+			glog.Errorf("CopyObjectHandler ValidateTags error %s: %v", r.URL, err)
+			s3err.WriteErrorResponse(w, r, s3err.ErrInvalidTag)
```
Comment (chrislusf) [inline weed/s3api/s3api_object_copy_handlers.go:51]: Is this tested at all? It should return here, right?

**Comment by lapshin-vitaly** [comment]
> @lapshin-vitaly Will you be able to add validation and copy tests in the same way ?
> https://github.com/chrislusf/seaweedfs/blob/master/weed/s3api/s3api_object_copy_handlers_test.go#L337

Add test case with invalid tag for copy and add tests for ValidateTags function

File: weed/s3api/s3api_object_copy_handlers.go (hunk 1 above)
Comment (lapshin-vitaly) [reply weed/s3api/s3api_object_copy_handlers.go:51]: yeah, it should return ErrInvalidTag error before put new file in storage

Comment (chrislusf) [reply weed/s3api/s3api_object_copy_handlers.go:51]: Need a "return" here.

File: weed/s3api/s3api_object_tagging_handlers.go [hunk 2]
```diff
@@ -62,22 +62,10 @@ func (s3a *S3ApiServer) PutObjectTaggingHandler(w http.ResponseWriter, r *http.R
 		return
 	}
 	tags := tagging.ToTags()
-	if len(tags) > 10 {
-		glog.Errorf("PutObjectTaggingHandler tags %s: %d tags more than 10", r.URL, len(tags))
+	err = validateTags(tags)
+	if err != nil {
+		glog.Errorf("PutObjectTaggingHandler ValidateTags error %s: %v", r.URL, err)
 		s3err.WriteErrorResponse(w, r, s3err.ErrInvalidTag)
```
Comment (chrislusf) [inline weed/s3api/s3api_object_tagging_handlers.go:68]: need to add a "return" here.

File: weed/s3api/tags.go [hunk 3]
```diff
@@ -37,3 +40,40 @@ func FromTags(tags map[string]string) (t *Tagging) {
...
+		if len(tag) == 2 {
+			parsedTags[tag[0]] = tag[1]
+		} else if len(tag) == 1 {
+			parsedTags[tag[0]] = ""
+		}
+	}
+	return nil, nil
+}
+
+func validateTags(tags map[string]string) error {
```
Comment (chrislusf) [inline weed/s3api/tags.go]: please add some tests for this.

File: weed/s3api/s3api_object_tagging_handlers.go (hunk 2 above)
Comment (lapshin-vitaly) [reply weed/s3api/s3api_object_tagging_handlers.go:68]: got it

File: weed/s3api/tags.go (hunk 3 above)
Comment (lapshin-vitaly) [reply weed/s3api/tags.go]: i add tests in this commit
https://github.com/chrislusf/seaweedfs/pull/3200/commits/e96937091301bea5483a729fe5a7a376f2ac64a4#diff-ab44d262d78b5074bf3fa6af9a98cfaf63fb661bf6b9261a66e77b9d0882294b

File: weed/s3api/s3api_object_copy_handlers.go (hunk 1 above)
Comment (lapshin-vitaly) [reply weed/s3api/s3api_object_copy_handlers.go:51]: got it

File: weed/s3api/s3api_object_copy_handlers_test.go [hunk 4]
```diff
@@ -332,14 +332,27 @@ var processMetadataBytesTestCases = []struct {
...
+	},
 }
 
 func TestProcessMetadata(t *testing.T) {
 	for _, tc := range processMetadataTestCases {
 		reqHeader := transferHToHeader(tc.request)
 		existing := transferHToHeader(tc.existing)
 		replaceMeta, replaceTagging := replaceDirective(reqHeader)
-
+		fmt.Println("test")
```
Comment (chrislusf) [inline weed/s3api/s3api_object_copy_handlers_test.go]: remove

Comment (lapshin-vitaly) [reply weed/s3api/s3api_object_copy_handlers_test.go]: got it

**Pull request merged**